---
title: "Bulk RNA-Seq Analysis of Control and Remyelinating hGPCs in vivo"
author: "John Mariani"
date: "1/30/2020"
output: 
  github_document:
   toc: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Read in RSEM gene output
```{R, echo = TRUE, message = FALSE, warning = FALSE}
library(tximport)
library(readr)
library(biomaRt)
library(EDASeq)
library(DESeq2)
library(RUVSeq)
library(ggfortify)
library(pheatmap)

#If not working with Raw RSEM genes.results, load RDS of tximport object

#temp = list.files(path = "data_for_import", pattern="genes.results")
#names(temp) <- substr(temp,1,nchar(temp)-14)
#txi.rsem <- tximport(paste0("data_for_import/",temp), type = "rsem")
#colnames(txi.rsem$abundance) <- names(temp)
#colnames(txi.rsem$counts) <- names(temp)
#colnames(txi.rsem$length) <- names(temp)
#saveRDS(txi.rsem, "data_for_import/tximport_RSEM.rds")
txi.rsem <- readRDS("data_for_import/tximport_RSEM.rds")
```

## Read in gene information from biomaRt

Grab gene information from biomaRt if you don't already have it in the RSEM_ folder

```{R}
#Create a dataframe of gene abundance estimates
TPM <- as.data.frame(txi.rsem$abundance)

ensemblFilename="data_for_import/ensemblGeneListH.csv"
if(file.exists(ensemblFilename)){
  ensemblGeneListH <- read.csv(ensemblFilename)} else{
    marth <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl", host = 'http://dec2017.archive.ensembl.org/')
    ensemblGeneListH <- getBM(attributes = c("ensembl_gene_id","external_gene_name", "gene_biotype", "description", "percentage_gene_gc_content"), filters = "ensembl_gene_id",values = row.names(txi.rsem$counts), mart = marth)
    write.csv(ensemblGeneListH, ensemblFilename)
  }
```

### Create  design and prep for EDA-Seq
```{R}
#Experimental design for Control (CTR) and 12 week post-cuprizone (CUP) CD140+ hGPC samples
sampleTableFull <- data.frame(group = factor(c(rep("Ctr.36",6),rep("Cup.36",6))))
rownames(sampleTableFull) <- colnames(txi.rsem$counts)

#Set 0 lengths to 1 as a DESeq2 requirement
txi.rsem$length[txi.rsem$length == 0] <- 1

#Annotate the abundance dataframe
TPM <- merge(TPM, ensemblGeneListH,by.x=0,by.y="ensembl_gene_id")

#Determine low expressed genes to filter out pre-analysis that may be due to technical noise
lowExpression <- data.frame(txi.rsem$counts[,1:2])
names(lowExpression) <- c("ctr36","cup36")
lowExpression$external <- TPM$external_gene_name
lowExpression$ctr36 <- rowMedians(txi.rsem$counts[,1:6])
lowExpression$cup36 <- rowMedians(txi.rsem$counts[,7:12])

#Filter out genes that are below a median of 3 estimated counts in both conditions
expressedGenes <- lowExpression[lowExpression$ctr36 > 3 | lowExpression$cup36 > 3,]
txi.rsem$counts <- txi.rsem$counts[row.names(txi.rsem$counts) %in% row.names(expressedGenes),]
txi.rsem$length <- txi.rsem$length[row.names(txi.rsem$length) %in% row.names(expressedGenes),]
txi.rsem$abundance <- txi.rsem$abundance[row.names(txi.rsem$abundance) %in% row.names(expressedGenes),]
ensemblGeneListH <- ensemblGeneListH[match(row.names(txi.rsem$counts), ensemblGeneListH$ensembl_gene_id),]
```

### GC Bias removal with EDASeq
```{R}
#Create dataframe to be used for EDASeq with sanity checks
uCovar <- data.frame(row.names = ensemblGeneListH$ensembl_gene_id, gccontent = ensemblGeneListH$percentage_gene_gc_content)
uCovar <- uCovar[row.names(uCovar) %in% row.names(txi.rsem$counts),, drop = F]
uCovar <- uCovar[match(row.names(txi.rsem$counts), row.names(uCovar)),, drop = F]

roundedCounts <- as.matrix(round(txi.rsem$counts,0))
eda <- newSeqExpressionSet(roundedCounts,featureData=uCovar, phenoData=data.frame(group = sampleTableFull$group, row.names = row.names(sampleTableFull)))
dataWithin <- withinLaneNormalization(eda,"gccontent", which="full")

#Replace RSEM object's counts with EDASeq normalized counts
txi.rsem$counts <- dataWithin@assayData$normalizedCounts
```

### Removal of Unwanted Variance via RUVSeq
```{R}
# Get Normalized Counts from DESeq2
dds36 <- DESeqDataSetFromTximport(txi.rsem, sampleTableFull, ~group)
dds36 <- estimateSizeFactors(dds36)
counts <- as.matrix(counts(dds36, normalized=TRUE))

#Generate RUVSeq covariate using RUVs
forRUVs <- newSeqExpressionSet(counts=as.matrix(round(counts,0)),featureData=uCovar, phenoData=data.frame(group = sampleTableFull$group, row.names = row.names(sampleTableFull)))
genes <- row.names(counts)
differences <- makeGroups(sampleTableFull$group)
s <- RUVs(forRUVs,genes,k=1,differences)
sampleDataW <- pData(s)
```

### Differential Expression in DESeq2
```{R}
ddsW <- DESeqDataSetFromTximport(txi.rsem, sampleDataW, design = ~group + W_1)
ddsW <- DESeq(ddsW, betaPrior = T)
deRUV <- function(cont){
  temp <- data.frame(results(ddsW, contrast =cont))
  temp <- merge(temp,ensemblGeneListH,by.x=0,by.y="ensembl_gene_id")
  return(temp[temp$padj < 0.05 & complete.cases(temp ==T),])
}


cupVsctrW <- deRUV(c("group","Cup.36", "Ctr.36"))

# Remove low abundance genes for further analysis
TPMhigh <- data.frame(row.names= TPM$Row.names, ctrMeans = rowMeans(TPM[,2:7]), ctrMedians = rowMedians(as.matrix(TPM[,2:7])), cupMeans = rowMeans(TPM[,8:13]), cupMedians = rowMedians(as.matrix(TPM[,8:13])))
TPMhigh <- merge(TPMhigh, ensemblGeneListH, by.x = 0, by.y = "ensembl_gene_id")
TPMhigh <- TPMhigh[TPMhigh$ctrMeans > 6.5 | TPMhigh$cupMeans > 6.5,]

#Filter out low abundance genes in DE list
cupVsctrWtpm <- cupVsctrW[cupVsctrW$Row.names %in% TPMhigh$Row.names,]
nrow(cupVsctrWtpm)
#write.csv(cupVsctrWtpm, cupVsctrWtpm)
```

### PCA
```{R}
#Make PCA from RUV Normalized Counts
pcaDF <- as.data.frame(log2(s@assayData$normalizedCounts+.5))

pcaDFTPM <- pcaDF[row.names(pcaDF) %in% TPMhigh$Row.names,]

condition_colors <- c(rep("blue",6),rep("red",6))
p <- autoplot(prcomp(t(pcaDFTPM)), xlim = c(-.5, 1.5), ylim = c(-1,.5), x = 1, y =2 , size = 2) 
p + theme_bw() + theme(panel.grid.major = element_line(colour = "grey")) + geom_point(colour = condition_colors, stroke =.2)
```
 
### Celltype Marker Heatmap
```{R}
CNSgenes <- read.csv("data_for_import/CNSgenesCup.csv")
pcaDF <- as.data.frame(log2(TPM[,2:13]+.5))
row.names(pcaDF) <- TPM$Row.names
pcaDFTPM <- pcaDF[row.names(pcaDF) %in% TPMhigh$Row.names,]

CNSgeneslog <- merge(pcaDF,ensemblGeneListH,by.x=0,by.y="ensembl_gene_id")
CNSgeneslog <- merge(CNSgenes,CNSgeneslog,by.x=1,by.y="external_gene_name")
CNSgeneslog <- CNSgeneslog[match(CNSgenes$Gene, CNSgeneslog$Gene),]
sigCNSlog <- data.frame(row.names = paste0(CNSgeneslog$Gene," (",CNSgeneslog$Lineage,")"), CNSgeneslog [,4:15])
my_palette <- colorRampPalette(c("#009900","#fffcbd","#ff2020"))(n=299)
pheatmap(sigCNSlog, border_color = "Black", cluster_row = FALSE, cluster_cols = FALSE, color = my_palette, labels_col = c(rep("",12)), gaps_col = 6)

```

### Make Functional and Module Organized Heatmaps of DE genes

```{R}
#Import genes for HMs labeled with relevant info
pathways <- read.csv("data_for_import/CupFinalPathways.csv")

pathwaysDF <- as.data.frame(pathways$Gene)
pathwaysDF <- merge(pathwaysDF, TPM, by.x = 1, by.y = "external_gene_name")
pathwaysDF <- data.frame(row.names = pathwaysDF$`pathways$Gene`, pathwaysDF[,3:14])
annotations <- data.frame(row.names = pathways$Gene, Module = pathways$Module)

mat_colors <- list(Module = c("#57bc9a", "#dd50dd", "#d78737", "#1b474a"))

breaks <- seq(from = -3, to = 3, length.out = 299)
makeHM <- function(category){
  temp <- TPM[TPM$external_gene_name %in% category$Gene,]
  temp <- data.frame(row.names = temp$external_gene_name, temp[,2:13])
  annotations <- data.frame(row.names = category$Gene, Module = category$Module)
  annotations <- annotations[order(annotations$Module,decreasing = F),, drop = F]
  temp <- temp[match(row.names(annotations), row.names(temp)),]
  pheatmap(log2(temp+.5), border_color = "Black", scale = "row", cluster_row = FALSE, cluster_cols = FALSE, color = my_palette,  labels_col = c(rep("",12)), gaps_col = 6, cellwidth = 10, cellheight = 10, annotation_row = annotations, annotation_colors = mat_colors, breaks = breaks, annotation_legend = T, legend = T, annotation_names_row = F, main = unique(category$Category))
}

adhesion <- pathways[pathways$Category == "Cell Adhesion and Structural Proteins",]
channels <- pathways[pathways$Category == "Channels and Transporters",]
enzymes <- pathways[pathways$Category == "Enzymes",]
kinases <- pathways[pathways$Category == "Kinases",]
ligands <- pathways[pathways$Category == "Ligands and Secreted Proteins",]
receptors <- pathways[pathways$Category == "Receptors and Downstream Components",]
tfs  <-  pathways[pathways$Category == "Transcriptional Regulators",]  

makeHM(adhesion)  
makeHM(channels) 
makeHM(enzymes)
makeHM(kinases)
makeHM(ligands)
makeHM(receptors)
makeHM(tfs)

```

### Make GO Bar plot
```{R}
### Import Module information and filter
GOterms <- read.delim("data_for_import/GOmodules.txt")
GOterms <- GOterms[GOterms$ForGraph == "Yes",]
GOterms$GO.Annotation <- droplevels(GOterms$GO.Annotation)
GOterms$GO.Annotation <- factor(GOterms$GO.Annotation, levels = rev(GOterms$GO.Annotation))
GOterms$Module <- factor(GOterms$Module)

attach(GOterms)
ggplot(GOterms, aes(fill=Module, x=GO.Annotation, y=X.log10.pvalue.)) + 
  geom_bar(position="dodge", stat="identity") + coord_flip() +
  ylab("Adjusted p-value (-log10)") + scale_y_continuous(expand = c(0, 0)) +
  xlab("GO Terms") + theme_minimal() + scale_fill_manual("Module", values = c("1" = mat_colors$Module[1], "2" = mat_colors$Module[2], "3" = mat_colors$Module[3], "4" = mat_colors$Module[4]))
```

### Session Info
```{R}
sessionInfo()
```